transport_nantes/mobilito/views.py

```diff
@@ -38,6 +38,132 @@ class TutorialView(TemplateView):
+            # convert it to a string.
+            "last_seen": self.now.strftime(self.strftime_pattern)
+        }
+
+        user = get_MobilitoUser(self)
+
+        # Next we look for pages that have been updated since the last time we
+        # visited them.
+        to_visit_pages = self.get_pages_to_visit(user)
+
+        if to_visit_pages and self.request.user.is_authenticated:
+            # The bottom arrow will use this arg to redirect to the next page.
+            context["the_next_page"] = to_visit_pages[0]
+            user.completed_tutorial_timestamp = None
+            user.save()
+        elif not to_visit_pages and self.request.user.is_authenticated:
+            logger.info(f"User {user} completed all tutorial pages")
+            user.completed_tutorial_timestamp = self.now
+            user.first_time = False
+            user.save()
+        elif to_visit_pages:
+            # The bottom arrow will use this arg to redirect to the next page.
+            context["the_next_page"] = to_visit_pages[0]
+
+        return context
+
+    def get_pages_to_visit(self, user: Union[MobilitoUser, None]) -> list:
+        """Return the list of pages to visit.
+
+        This is used by the bottom arrow to redirect to the next page.
+        """
+        to_visit_pages = []
+
+        for tutorial in self.all_tutorial_pages:
+
+            if (user and user.completed_tutorial_timestamp
+                    and
+                    user.completed_tutorial_timestamp
+                    < self.all_tutorial_pages[tutorial]["last_modified"]):
+                # The page has been updated since we saved that the user completed
+                # the tutorial, we can skip additional checks.
+                # This is mostly a backup if the user happen to delete cookies.
+                to_visit_pages.append(tutorial)
+                continue
+
+            # The user has already seen this page.
+            if tutorial in self.request.session:
+                # We recreate a datetime object from the string we stored in the
+                # session. Because in the session we stored it as a string,
+                # we lost the timezone information, so we need to add it back.
+                last_seen = datetime.strptime(
+                    self.request.session[tutorial]["last_seen"],
+                    self.strftime_pattern).replace(tzinfo=timezone.utc)
+                # The page has been updated since the last time we saw it.
+                if self.all_tutorial_pages[tutorial]["last_modified"] > last_seen:
+                    # The tutorial page exists, and has been updated since the
+                    # last time the user saw it, according to session cookie
+                    to_visit_pages.append(tutorial)
+                # The page has not been updated since the last time we saw it.
+                # But it's been more than 60 days since they saw it.
+                elif (self.now - last_seen).days > 60:
```

Comment:
Please drop this entire block.

1. We have no data to support its necessity.
2. I rather suspect marketing emails will be a better way to do it anyway.
3. In so much as we haven't yet thought through the user experience, and that will require extra work, let's not put it on the critical path to getting this code delivered, since if it's here, we have to test it and provide reasonable user experience for it.
4. Why should "60" be hard-coded as a magic number?

Reply:
Dropped in next push